xrRender: static geometry buffers code unification (#489)

vTurbine opened the pull request:
- Added staging buffer wrapper for static geometry;
- Common buffer interface for DX9/DX10/DX11/VK
- Backend top file simplified and made common

Please check if it works fine on Linux as well: I don't have OpenGL 4.x on my VM

vTurbine committed:
xrRender: added index staging buffer abstraction

vTurbine committed:
xrRender: Backend: top file cleanup

vTurbine committed:
xrRender: added vertex staging buffer abstraction

vTurbine committed:
xrRender: Details: detail manager switched to staging buffers

vTurbine committed:
xrRender: vertex format helpers interface unified

vTurbine committed:
xrRender: Utils: geometry switched to staging buffers

vTurbine committed:
xrRender: BufferUtils: staging buffer validation added

vTurbine committed:
xrRender: BufferUtils: staging buffer is reusable now

vTurbine committed:
xrRender: renderers were switched to common `BufferUtils.h`

vTurbine committed:
xrRenderR2: loader switched to staging geometry buffers

vTurbine committed:
xrRenderR1: loader switched to staging geometry buffers

vTurbine committed:
xrRender: BufferUtils: added ref counting into staging buffer

vTurbine committed:
xrRenderR2: RT switched to staging buffers

vTurbine committed:
xrRenderR3: loader switched to staging geometry buffers

Xottab-DUTY commented on src/Layers/xrRenderDX9/CommonTypes.h:
```diff
@@ -20,3 +20,12 @@ typedef IDirect3DVolumeTexture9 ID3DTexture3D;
 
 #define DX10_ONLY(expr) do {} while (0)
+
+using Unused = int[0];
```
```suggestion
using unused_t = int[0];
```

SkyLoaderr commented on src/Layers/xrRenderDX10/dx10BufferUtils.cpp:
```diff
@@ -145,3 +145,37 @@ void ConvertVertexDeclaration(const xr_vector<D3DVERTEXELEMENT9>& declIn, xr_vec
+    HW.stats_manager.increment_stats_ib(m_DeviceBuffer);
+    // Free host memory
+    xr_delete(m_HostData);
+    m_HostData = nullptr;
```
xr_delete already has `ptr = nullptr`

vTurbine committed:
xrRenderR4: loader switched to staging buffers

SkyLoaderr commented on src/Layers/xrRenderPC_R2/r2_loader.cpp:
```diff
@@ -264,13 +258,11 @@ void CRender::LoadBuffers(CStreamReader* base_fs, bool alternative)
-            R_CHK(HW.pDevice->CreateIndexBuffer(iCount * 2, dwUsage, D3DFMT_INDEX16, D3DPOOL_MANAGED, &_IB[i], 0));
-            HW.stats_manager.increment_stats_ib(_IB[i]);
-            R_CHK(_IB[i]->Lock(0, 0, (void**)&pData, 0));
+            _IB[i].Create(iCount * 2);
```
Old code doesn't use `D3DUSAGE_SOFTWAREPROCESSING` for `dwUsage` in software mode, but `_IB[i].Create` has it.

vTurbine replied:
This odd for level geometry loading only. I believe this typo came from past times and the flag should be set. See R1 for reference.
Any reasons to treat this particular geometry buffers in different manner? Am I missing smth?

Xottab-DUTY replied:
> This odd for level geometry loading only. I believe this typo came from past times and the flag should be set. See R1 for reference.
Any reasons to treat this particular geometry buffers in different manner? Am I missing smth?

@SkyLoaderr, so, do you have something to say about it?

Xottab-DUTY replied:
I bet this is not a typo and was made intentional. But, however, R2 probably won't even launch without hardware features, so this should be safe.

SkyLoaderr commented on src/Layers/xrRender/BufferUtils.h:
```diff
@@ -31,6 +31,19 @@ class VertexStagingBuffer
+    u32 Release()
+    {
+        VERIFY(m_RefCounter); // attempting to release unused object
+        if (--m_RefCounter == 0)
```
Maybe separate these two operations?

Xottab-DUTY commented on src/Layers/xrRender/BufferUtils.h:
```diff
@@ -0,0 +1,114 @@
+
+    u32 Release()
+    {
+        VERIFY(m_RefCounter); // attempting to release unused object
```
```suggestion
        VERIFY2(m_RefCounter, "Attempt to release unused object");
```

vTurbine committed:
Review iteration

Xottab-DUTY commented on src/Layers/xrRenderPC_R2/r2_rendertarget_accum_point_geom.cpp:
```diff
@@ -2,328 +2,38 @@
 #include "Layers/xrRender/du_sphere.h"
 
-/*
-Fvector du_sphere_vertices[DU_SPHERE_NUMVERTEX]=
```
Please, don't remove this, because dead code cleanup and this code itself is not related to this PR.

Xottab-DUTY commented on src/Layers/xrRenderPC_R2/r2_rendertarget_accum_spot_geom.cpp:
```diff
@@ -1,107 +1,41 @@
 
 #include "Layers/xrRender/du_cone.h"
-/*
-Fvector du_cone_vertices[DU_CONE_NUMVERTEX]=
```
Please, don't remove this, because dead code cleanup and this code itself is not related to this PR.

Xottab-DUTY commented on src/Layers/xrRender/BufferUtils.h:
```diff
@@ -0,0 +1,116 @@
+    void Destroy();
+
+    bool IsValid() const;
+    void* GetHostPointer();
```
It is better to rename this function to `Lock`.
```suggestion
    void* Lock();
```

SkyLoaderr replied:
I agree

vTurbine replied:
Changed to `Map`/`Unmap` pair to be consistent with forthcoming streaming buffers

Xottab-DUTY replied:
It's even more complicated now and should be made easier.

Xottab-DUTY replied:
It is absolutely obvious to call `Unlock()` if you called `Lock()`.
It is less, but obvious to call `Unmap()` if you called `Map()`.

But it is non-obvious to call `Flush()` if you called `Map()`

vTurbine replied:
This complexity was added with purpose. In the case of skinned meshes loading, we will need access to buffer data again.
I can add all the code reworked but this will increase the PR. Or can share my local branch with all geometry stuff reworked

vTurbine replied:
Agree with Map/Unmap inconsistency

Xottab-DUTY replied:
> This complexity was added with purpose. In the case of skinned meshes loading, we will need access to buffer data again.
I can add all the code reworked but this will increase the PR. Or can share my local branch with all geometry stuff reworked

Then we should remove `Unmap()` function and rename `Flush()` to `FlushAndUnmap()`. This is very consistent:
1. Linked with `Map()` functions (function name tells you that you should call it)
2. Corresponds with what the functions does on DX10+/OpenGL. (It flushes the buffer and then discards the host buffer)

Xottab-DUTY commented on src/Layers/xrRender/BufferUtils.h:
```diff
@@ -0,0 +1,116 @@
+    bool IsValid() const;
+    void* GetHostPointer();
+    VertexBufferHandle GetBufferHandle() const;
+    void Flush();
```
It is better to rename this function to `Unlock`.
```suggestion
    void Unlock();
```

Xottab-DUTY commented on src/Layers/xrRender/BufferUtils.h:
```diff
@@ -0,0 +1,116 @@
+    bool IsValid() const;
+    void* GetHostPointer();
+    IndexBufferHandle GetBufferHandle() const;
+    void Flush();
```
It is better to rename this function to `Unlock`.
```suggestion
    void Unlock();
```

Xottab-DUTY commented on src/Layers/xrRender/BufferUtils.h:
```diff
@@ -0,0 +1,116 @@
+    void Destroy();
+
+    bool IsValid() const;
+    void* GetHostPointer();
```
It is better to rename this function to `Lock`.
```suggestion
    void* Lock();
```

vTurbine committed:
Review iteration #2

* Dead code restored
* `GetHostPointer` -> `Map`
* Option to keep host buffer for further readings
* `Destroy` is hidden to prevent ref counter miscalculations

vTurbine committed:
Assert statement simplified

vTurbine committed:
Merge branch 'xd_dev' into renderer/staging_buffers

vTurbine committed:
Merge issue resolved

vTurbine committed:
Review iteration #3

- `Flush` removed;
- More verbosity in assertions;
- Don't allow to map buffer for writes twice. Such pattern isn't used
in the Engine and this will need buffer handle invalidation.
- Some documentation added

Xottab-DUTY commented on src/Layers/xrRenderGL/glBufferUtils.cpp:
```diff
@@ -278,4 +280,197 @@ u32 GetDeclLength(const D3DVERTEXELEMENT9* decl)
+    if (!m_DeviceBuffer)
+        return 0;
+
+    GLint VB_size;
```
Local variables should be in camelCase.

vTurbine replied:
Yeah, silly copy-paste issue

vTurbine committed:
Coding style fixed

Xottab-DUTY reviewed (approved):
@vTurbine, anything more to apply or we may merge?

vTurbine committed:
`r_restore_quad_ib_data` removed
